events/emitter.go

```diff
@@ -0,0 +1,159 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/api/types/event"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/gogo/protobuf/proto"
+	protobuf "github.com/gogo/protobuf/types"
+)
+
+const (
+	EventVersion = "v1"
+)
+
+type listener struct {
+	ns string
+	ch chan *event.Envelope
+}
+
+type Emitter struct {
+	listeners map[string]*listener
```

Comment:
Isn't some synchronization necessary on the various methods that use `listeners`? It looks like these are invoked from different goroutines.